components/org.wso2.carbon.aws/src/main/java/org/wso2/carbon/aws/user/store/mgt/util/AWSSignatureV4Generator.java

```diff
@@ -0,0 +1,424 @@
+
+            // Step 3.2.1 Encode signature (byte[]) to Hex.
+            return bytesToHex(signature);
+        } catch (UnsupportedEncodingException | InvalidKeyException | NoSuchAlgorithmException ex) {
+            log.error("Error while calculating the signature." + ex);
+        }
+        return null;
+    }
+
+    /**
+     * Task 4: Add the Signing Information to the Request. We'll return Map of
+     * all headers put this headers in your request.
+     *
+     * @return Headers.
+     */
+    public Map<String, String> getHeaders() {
+
+        awsHeaders.put(AWSConstants.DATE_HEADER, xAmzDate);
+
+        // Execute Task 1: Create a Canonical Request for Signature Version 4.
+        String canonicalURL = prepareCanonicalRequest();
+
+        // Execute Task 2: Create a String to Sign for Signature Version 4.
+        String stringToSign = prepareStringToSign(canonicalURL);
+
+        // Execute Task 3: Calculate the AWS Signature Version 4.
+        String signature = calculateSignature(stringToSign);
+
+        if (signature != null) {
+            Map<String, String> header = new HashMap<>(0);
+            header.put(AWSConstants.DATE_HEADER, xAmzDate);
+            header.put(AWSConstants.AUTHORIZATION_HEADER, buildAuthorizationString(signature));
+
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Signature: %s. Headers: ", signature));
+                for (Map.Entry<String, String> entrySet : header.entrySet()) {
+                    log.debug(entrySet.getKey() + " = " + entrySet.getValue());
+                }
+            }
+            return header;
+        }
+        return null;
+    }
+
+    /**
+     * Build string for Authorization header.
+     *
+     * @param strSignature Signature value.
+     * @return Authorization String.
+     */
+    private String buildAuthorizationString(String strSignature) {
+
+        return AWSConstants.HMAC_ALGORITHM + " "
+                + "Credential=" + accessKeyID + "/" + getDate() + "/" + regionName + "/" + serviceName + "/"
+                + AWSConstants.AWS4_REQUEST + ","
+                + "SignedHeaders=" + signedHeaderString + ","
+                + "Signature=" + strSignature;
+    }
+
+    /**
+     * Hashes the string contents (assumed to be UTF-8) using the SHA-256 algorithm.
+     *
+     * @param data text to be hashed
+     * @return SHA-256 hashed text
+     */
+    private String hash(String data) {
+
+        MessageDigest messageDigest;
+        try {
+            messageDigest = MessageDigest.getInstance(AWSConstants.SHA_ALGORITHM);
+            messageDigest.update(data.getBytes(AWSConstants.UTF_8));
+            byte[] digest = messageDigest.digest();
+            return String.format("%064x", new java.math.BigInteger(1, digest));
+        } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
+            log.error("Error while hashing the string contents." + e);
```

Comment:
Why do we need to handle the exception here and return a null ? Cant we re-throw or wrap and throw ? and then decide what the next steps are based on exception ?